cluster/cluster_interface.ml

```diff
@@ -0,0 +1,214 @@
+   "if None"]]
+[@@deriving rpcty]
+
+type cluster_config = {
+  cluster_name : string;
+  enabled_members : node list;
+  authkey: string;
+  config_version: int64;
+  cluster_token_timeout_ms : int64;
+  cluster_token_coefficient_ms : int64;
+}
+[@@doc
+  ["This type contains all of the information required to configure";
+   "the cluster. This includes all details required for the corosync";
+   "configuration as well as anything else required for pacemaker and";
+   "SBD. All nodes have a local copy of this and we take pains to";
+   "ensure it is kept in sync."]]
+[@@deriving rpcty]
+
+type cluster_config_and_all_members = cluster_config * all_members [@@deriving rpcty]
+
+type diagnostics = {
+  config_valid : bool;
+  live_cluster_config : cluster_config option; (* live corosync config *)
+  next_cluster_config : cluster_config option; (* next corosync config *)
+  saved_cluster_config : cluster_config option; (* xapi-clusterd DB *)
+  is_enabled : bool;
+  all_members : all_members option;
+  node_id : nodeid option;
+  token : string option;
+  num_times_booted : int;
+  is_quorate : bool;
+  is_running : bool;
+}
+[@@doc
+ [ "This type contains diagnostic information about the current state";
+   "of the cluster daemon. All state required for test purposes should";
+   "be in this type."]]
+[@@deriving rpcty]
+
+type token = string
+[@@doc ["This secret token is used to authenticate remote API calls on a cluster"]]
+[@@deriving rpcty]
+
+let token_p = Param.mk ~name:"token" token
+
+type error =
+  | InternalError of string
+  | Unix_error of string
+[@@deriving rpcty]
+
+module E = Error.Make(struct
+    type t = error [@@deriving rpcty]
+  end)
+let err = E.error
+
+type named_unit = unit [@@deriving rpcty]
+type my_string = string [@@deriving rpcty]
+
+
+let unit_p         = Param.mk ~name:"unit" ~description:["unit"] named_unit
+let string_p       = Param.mk ~name:"string" ~description:["string"] my_string
+let address_p      = Param.mk ~name:"address" ~description:[
+    "IP address of a cluster member";
+  ] address
+let init_config_p     = Param.mk ~name:"init_config" ~description:[
+    "The initial config of the cluster member";
+  ] init_config
+
+let debug_info_p = Param.mk ~name:"dbg" ~description:[
+                       "An uninterpreted string to associate with the operation."
+                     ] debug_info
+
+type remove = bool [@@deriving rpcty]
+
+module LocalAPI(R:RPC) = struct
+  open R
+
+  let description = Interface.{
+      name = "Local";
+      namespace = None;
+      description = [
+        "Local Cluster APIs. These are intended to be used to control the xapi-clusterd service";
+        "There is no authentication on these, but they are only available on the local machine.";
+      ];
+      version = (1,0,0);
+    }
+
+  let implementation = implement description
+
+  let create = declare
+      "create"
+      ["Creates the cluster. The call takes the initial config of";
```

Comment:
When we concatenate these strings, we get something like ".. of;the ..". Is this intended?

Reply:
The `;` won't be part of the concatenation, it separates the strings in the ocaml list.

Reply:
That's right. So the concatenation would add a space?

Reply:
Yes it concatenates the strings with spaces: https://github.com/mirage/ocaml-rpc/blob/474d9cd1a10de4f6078a6422e14a91ed84d8173b/lib/markdowngen.ml#L97